### PR TITLE
[Calling] Feature : Active speaker - part 2

### PR DESCRIPTION
**Pull request by mejdoo**
This is a follow up for the calling feature **Active Speaker Detection** introduced in #3106 

## What's new ?

- Highlight the top 4 active speakers having their videos enabled during a conference call.
- Switching between All participants and active speakers using a toggle in call controls.
- Manage the case when there is no active speakers

Design spec : https://zpl.io/bAoEQeP
#### APK
[Download build #3025](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3025/artifact/build/artifact/wire-dev-PR3122-3025.apk)
[Download build #3031](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3031/artifact/build/artifact/wire-dev-PR3122-3031.apk)
[Download build #3033](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3033/artifact/build/artifact/wire-dev-PR3122-3033.apk)

### CODE REVIEW

File: app/src/main/scala/com/waz/zclient/calling/controllers/CallController.scala [hunk 1]
```diff
@@ -113,6 +114,31 @@ class CallController(implicit inj: Injector, cxt: WireContext)
...
+
+        var activeParticipantsWithVideo: Set[Participant] = Set.empty
+
+        activeSpeakers.foreach { speaker =>
+          videoUsers.foreach { participant =>
+            if (participant.clientId == speaker.clientId && participant.userId == speaker.userId)
+              activeParticipantsWithVideo = activeParticipantsWithVideo + participant
+          }
+        }
+        activeParticipantsWithVideo.toSeq
```
Comment (makingthematrix) [inline app/src/main/scala/com/waz/zclient/calling/controllers/CallController.scala:130]: Try this instead:
```
videoUsers.filter { participant => 
  activeSpeakers.exists { speaker => 
    participant.clientId == speaker.clientId && participant.userId == speaker.userId
  }
}
```

File: app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala [hunk 2]
```diff
@@ -141,50 +151,61 @@ class CallingFragment extends FragmentHelper {
...
-      case _: SelfVideoView if views.size > 1 && !isVideoBeingSent => false
-      case _ => true
-    }.sortWith {
-      case (_: SelfVideoView, _) => true
-      case (_, _: SelfVideoView) => false
-      case (v1, v2) =>
-        infoMap(v1.participant.userId).displayName.toLowerCase < infoMap(v2.participant.userId).displayName.toLowerCase
-    }.take(MaxVideoPreviews)
+    var gridViews: Seq[UserVideoView] = Seq.empty
+
```
Comment (makingthematrix) [inline app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala:190]: You can turn this `var` into a val and assign the result of if/else to it:
```
val gridViews = 
  if (showTopSpeakers) 
    views.sortWith {
      ..
    }.take(MaxTopSpeakerVideoPreviews)
  else 
    views.filter {
      ...
    }.sortWith {
      ...
    }.take(MaxAllVideoPreviews)
```

File: app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala [hunk 3]
```diff
@@ -224,8 +245,8 @@ class CallingFragment extends FragmentHelper {
     viewMap = viewMap.filter { case (participant, _) => videoUsers.contains(participant) }
   }
 
-  def clearVideoGrid(): Unit = {
-    videoGrid.foreach(_.removeAllViews())
+  def clearVideoGrids(): Unit = {
+    videoGrid.foreach (_.removeAllViews())
```
Comment (makingthematrix) [inline app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala:249]: Usually it's `.foreach(...)` without a whitespace and `.foreach { ... }` with a whitespace.

File: app/src/main/scala/com/waz/zclient/calling/ControlsFragment.scala [hunk 4]
```diff
@@ -73,6 +84,12 @@ class ControlsFragment extends FragmentHelper {
         getContext.startActivity(new Intent(getContext, classOf[MainActivity]))
       }
     }
+
+    Signal.zip(controller.isCallEstablished, controller.isGroupCall){
```
Comment (makingthematrix) [inline app/src/main/scala/com/waz/zclient/calling/ControlsFragment.scala:88]: This should be `Signal.zip(controller.isCallEstablished, controller.isGroupCall).foreach {`.
If you write it without `.foreach` it means you are using the `apply` method instead and it works a bit differently... although there shouldn't be any difference in this case (the `apply` method subscribes to the signal without specifying the execution context, but here we use `Threading.Ui` as the execution context everywhere so there's no risk to mix it up). 

I think I will remove this `apply` from `wire-signals` soon :)

**Review by makingthematrix — CHANGES_REQUESTED** [review]

File: app/src/main/scala/com/waz/zclient/calling/ControlsFragment.scala (hunk 4 above)
Comment (mejdoo) [reply app/src/main/scala/com/waz/zclient/calling/ControlsFragment.scala:88]: ok done

File: app/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -336,6 +336,7 @@
     <string name="group_calling_message" tools:ignore="PluralsCandidate">This will call %1$s people</string>
     <string name="group_calling_confirm">Call</string>
     <string name="calling_self">%s (You)</string>
+    <string name="calling_no_active_speaker">No active speakers…</string>
```
[the remark this answers is not in the document]
Comment (mejdoo) [reply app/src/main/res/values/strings.xml:337]: Sorry, I didn't get what you mean here.

**Review by makingthematrix — APPROVED** [review]

**Review by makingthematrix — APPROVED** [review]

**Review by gizemb — APPROVED** [review]

File: app/src/main/res/layout/fragment_calling_controls.xml [hunk 6]
```diff
@@ -42,6 +42,16 @@
         android:layout_marginBottom="@dimen/wire__padding__32"
         android:layout_weight="1" />
 
+    <ToggleButton
+        android:id="@+id/all_speakers_toggle"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:textOff="ALL"
+        android:textOn="SPEAKERS"
```
[the remark this answers is not in the document]
Comment (mejdoo) [reply app/src/main/res/layout/fragment_calling_controls.xml:50]: right, I forgot to do it. Thanks for the heads up!

**Pull request merged**